scylla/src/transport/connection.rs

```diff
@@ -4,6 +4,7 @@ use scylla_cql::errors::TranslationError;
 use scylla_cql::frame::request::options::Options;
 use scylla_cql::frame::response::Error;
 use scylla_cql::frame::types::SerialConsistency;
+use scylla_cql::frame::value::SerializedValues;
```

Comment:
> connection.rs: Don't allow queries with values
>
> After serialization refactor it will be impossible to perform unprepared
> query with values, because serializing values will require knowing
> column types.
> 
> In order to make refactor easier and better split responsibility, this
> commit removes `values` arguments from `Connection` methods, so that
> it is callers responsibility to prepare the query if necessary.

I'm not sure if I see why it makes the remaining refactor easier. It looks like you are simplifying the API of `Connection` but this is just pushing this concern to the caller. This requires modifying lots of places in the code, including tests and `query_filter_keyspace_name` which manually checks whether values is empty and dispatches to `query` or `execute` anyway.

I wonder whether it wouldn't be just simpler to drop this commit and then rework the last one, but I may be missing something.

Reply:
My thinking is that we don't want to send values in unprepared queries, and removing such capability from `Connection` is a way to guarantee that. If the only users of this functionality are tests and `query_filter_keyspace_name`, then I think it's not worth it to have this functionality.

I'll look into this matter again - I remember that some changes were less pleasant than I hoped, maybe reverting those changes in connections would help with that.

Reply:
> My thinking is that we don't want to send values in unprepared queries, and removing such capability from `Connection` is a way to guarantee that.

What do you mean exactly by "send values in unprepared queries"? We are no longer doing that on the protocol level after https://github.com/scylladb/scylla-rust-driver/pull/806 and https://github.com/scylladb/scylla-rust-driver/pull/812 . If you mean to disallow calling `.query()` with values, then I'm not sure if it makes sense to disallow it for `Connection` but not for `Session` (and AFAIK we didn't decide to do it for `Session` at this point).

Reply:
Removing this commit opens us up to the problems like following:
`topology::query_filter_keyspace_name` calls `Connection::query_iter`, potentially with values.
`Connection::query_iter` will call `RowIterator::new_for_connection_query_iter`, which will,
for each page, call `Connection::query_with_consistency` - which, if there are values,
prepares the query underneath to serialize them.
Such problems seem pretty hard to notice normally but removing `values` from Connection pretty much prevents them.

Currently, I tried to revert removal of `values` from `Connection` and from `RowIterator::new_for_query`, as it forced me to prepare a query globally in `Session::query_iter` if there are values - now it is again prepared for each sub query.
Doing this created another problem I'm not yet sure how to solve: - if `RowIterator::new_for_query` accepts `values`, how do we get a size of it
for request span? In either variant of `Connection` API, `new_for_query` won't serialize the values, `Connection` will do it. There is no place where `RowIterator` sees serialized values, so it can't record it's size for span.

For a user of `Connection` struct (e.g. struct `Session`), API without `values` args covers all the use cases that the previous one did - perhaps with a bit more code sometimes, as the caller needs to prepare the query explicitly if they want to send values.

One advantage I see of removing `values` from `Connection` is that it makes the API more explicit - additional overhead of preparing queries is not hidden underneath `Connection::query*` methods, so we are more aware of it while writing other code.

Reply:
> Removing this commit opens us up to the problems like following: `topology::query_filter_keyspace_name` calls `Connection::query_iter`, potentially with values. `Connection::query_iter` will call `RowIterator::new_for_connection_query_iter`, which will, for each page, call `Connection::query_with_consistency` - which, if there are values, prepares the query underneath to serialize them. Such problems seem pretty hard to notice normally but removing `values` from Connection pretty much prevents them.

This issue is easy to fix - prepare the query if needed in `Connection::query_iter` and then change `RowIterator::new_connection_query_iter` to be `RowIterator::new_connection_execute_iter`.

> Currently, I tried to revert removal of `values` from `Connection` and from `RowIterator::new_for_query`, as it forced me to prepare a query globally in `Session::query_iter` if there are values - now it is again prepared for each sub query.

This sounds like a harder problem because, unlike `Connection::query_iter`, the iterator might switch to another connection when performing a query fails. Perhaps we could be smarter here and prepare the query when `RowIteratorWorker::work` chooses a connection, but doing that sounds like too much work to be worth it, so your current approach looks OK for this one.

> Doing this created another problem I'm not yet sure how to solve: - if `RowIterator::new_for_query` accepts `values`, how do we get a size of it for request span? In either variant of `Connection` API, `new_for_query` won't serialize the values, `Connection` will do it. There is no place where `RowIterator` sees serialized values, so it can't record it's size for span.

This problem should go away if you apply my first suggestion.

> For a user of `Connection` struct (e.g. struct `Session`), API without `values` args covers all the use cases that the previous one did - perhaps with a bit more code sometimes, as the caller needs to prepare the query explicitly if they want to send values.
> 
> One advantage I see of removing `values` from `Connection` is that it makes the API more explicit - additional overhead of preparing queries is not hidden underneath `Connection::query*` methods, so we are more aware of it while writing other code.

You can look at it from another angle - having to do manual prepare in metadata fetch code sounds like polluting it with unnecessary details which should be handled on a lower level.

----

The examples with `query_iter` convince me that it might be a good idea after all to automatically prepare on the highest layer instead on the lowest, because otherwise unnecessary preparing may occur. Though I'd prefer that `query_iter` API did prepare automatically like I described (I see it as a user API in some sense, where the control connection is the user), there is an issue about changing internal CQL statements to use prepared statements (https://github.com/scylladb/scylla-rust-driver/issues/417) so I suppose having `Connection::execute_iter` right now is acceptable.

To sum up - let's not pursue my original suggestion from my first comment in this thread, I don't think it's worth it anymore.